File: src/demos/irrlicht/demo_IRR_ujoint.cpp
// =============================================================================
// PROJECT CHRONO - http://projectchrono.org
//
// Copyright (c) 2014 projectchrono.org
// All rights reserved.
//
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file at the top level of the distribution and at
// http://projectchrono.org/license-chrono.txt.
//
// =============================================================================
// Authors: Radu Serban
// =============================================================================
//
// Demo for universal joint.
//
// Recall that Irrlicht uses a left-hand frame, so everything is rendered with
// left and right flipped.
//
// =============================================================================

#include <cmath>

#include "chrono/physics/ChBody.h"
#include "chrono/physics/ChLinkMotorRotationAngle.h"
#include "chrono/physics/ChSystemNSC.h"

#include "chrono_irrlicht/ChIrrApp.h"

using namespace chrono;
using namespace chrono::irrlicht;

using namespace irr;

// -----------------------------------------------------------------------------

int main(int argc, char* argv[]) {
    GetLog() << "Copyright (c) 2017 projectchrono.org\nChrono version: " << CHRONO_VERSION << "\n\n";

    ChSystemNSC system;

    // Disable gravity
    system.Set_G_acc(ChVector<>(0, 0, 0));

    // Set the half-length of the two shafts
    double hl = 2;

    // Set the bend angle between the two shafts (positive rotation about the
    // global X axis)
    double angle = CH_C_PI / 6;
    double cosa = std::cos(angle);
    double sina = std::sin(angle);
    ChQuaternion<> rot = Q_from_AngX(angle);

    // Create the ground (fixed) body
    // ------------------------------

    auto ground = std::make_shared<ChBody>();
    system.AddBody(ground);
    ground->SetIdentifier(-1);
    ground->SetBodyFixed(true);
    ground->SetCollide(false);

    // attach visualization assets to represent the revolute and cylindrical
    // joints that connect the two shafts to ground
    {
        auto cyl_1 = std::make_shared<ChCylinderShape>();
        cyl_1->GetCylinderGeometry().p1 = ChVector<>(0, 0, -hl - 0.2);
        cyl_1->GetCylinderGeometry().p2 = ChVector<>(0, 0, -hl + 0.2);
        cyl_1->GetCylinderGeometry().rad = 0.3;
        ground->AddAsset(cyl_1);

        auto cyl_2 = std::make_shared<ChCylinderShape>();
        cyl_2->GetCylinderGeometry().p1 = ChVector<>(0, -(hl - 0.2) * sina, (hl - 0.2) * cosa);
        cyl_2->GetCylinderGeometry().p2 = ChVector<>(0, -(hl + 0.2) * sina, (hl + 0.2) * cosa);
        cyl_2->GetCylinderGeometry().rad = 0.3;
        ground->AddAsset(cyl_2);
    }

    // Create the first shaft body
    // ---------------------------

    auto shaft_1 = std::make_shared<ChBody>();
    system.AddBody(shaft_1);
    shaft_1->SetIdentifier(1);
    shaft_1->SetBodyFixed(false);
    shaft_1->SetCollide(false);
    shaft_1->SetMass(1);
    shaft_1->SetInertiaXX(ChVector<>(1, 1, 0.2));
    shaft_1->SetPos(ChVector<>(0, 0, -hl));
    shaft_1->SetRot(ChQuaternion<>(1, 0, 0, 0));

    // Add visualization assets to represent the shaft (a box) and the arm of the
    // universal joint's cross associated with this shaft (a cylinder)
    {
        auto box_1 = std::make_shared<ChBoxShape>();
        box_1->GetBoxGeometry().Size = ChVector<>(0.15, 0.15, 0.9 * hl);
        shaft_1->AddAsset(box_1);

        auto cyl_2 = std::make_shared<ChCylinderShape>();
        cyl_2->GetCylinderGeometry().p1 = ChVector<>(-0.2, 0, hl);
        cyl_2->GetCylinderGeometry().p2 = ChVector<>(0.2, 0, hl);
        cyl_2->GetCylinderGeometry().rad = 0.05;
        shaft_1->AddAsset(cyl_2);

        auto col = std::make_shared<ChColorAsset>();
        col->SetColor(ChColor(0.6f, 0, 0));
        shaft_1->AddAsset(col);
    }

    // Create the second shaft body
    // ----------------------------

    // The second shaft is identical to the first one, but initialized at an angle
    // equal to the specified bend angle.

    auto shaft_2 = std::make_shared<ChBody>();
    system.AddBody(shaft_2);
    shaft_2->SetIdentifier(1);
    shaft_2->SetBodyFixed(false);
    shaft_2->SetCollide(false);
    shaft_2->SetMass(1);
    shaft_2->SetInertiaXX(ChVector<>(1, 1, 0.2));
    shaft_2->SetPos(ChVector<>(0, -hl * sina, hl * cosa));
    shaft_2->SetRot(rot);

    // Add visualization assets to represent the shaft (a box) and the arm of the
    // universal joint's cross associated with this shaft (a cylinder)
    {
        auto box_1 = std::make_shared<ChBoxShape>();
        box_1->GetBoxGeometry().Size = ChVector<>(0.15, 0.15, 0.9 * hl);
        shaft_2->AddAsset(box_1);

        auto cyl_2 = std::make_shared<ChCylinderShape>();
        cyl_2->GetCylinderGeometry().p1 = ChVector<>(0, -0.2, -hl);
        cyl_2->GetCylinderGeometry().p2 = ChVector<>(0, 0.2, -hl);
        cyl_2->GetCylinderGeometry().rad = 0.05;
        shaft_2->AddAsset(cyl_2);

        auto col = std::make_shared<ChColorAsset>();
        col->SetColor(ChColor(0, 0, 0.6f));
        shaft_2->AddAsset(col);
    }

    // Connect the first shaft to ground
    // ---------------------------------

    // Use a rotational motor to impose both the revolute joint constraints, as well
    // as constant angular velocity. Here, we drive the motor angle with a ramp function.
    // Alternatively, we could use a ChLinkMotorAngularSpeed with constant speed.
    // The joint is located at the origin of the first shaft.
    auto motor = std::make_shared<ChLinkMotorRotationAngle>();
    motor->Initialize(ground, shaft_1, ChFrame<>(ChVector<>(0, 0, -hl), ChQuaternion<>(1, 0, 0, 0)));
    motor->SetAngleFunction(std::make_shared<ChFunction_Ramp>(0, 1));
    system.AddLink(motor);

    // Connect the second shaft to ground through a cylindrical joint
    // --------------------------------------------------------------

    // Use a cylindrical joint so that we do not have redundant constraints
    // (note that, technically Chrono could deal with a revolute joint here).
    // the joint is located at the origin of the second shaft.

    auto cyljoint = std::make_shared<ChLinkLockCylindrical>();
    system.AddLink(cyljoint);
    cyljoint->Initialize(ground, shaft_2, ChCoordsys<>(ChVector<>(0, -hl * sina, hl * cosa), rot));

    // Connect the two shafts through a universal joint
    // ------------------------------------------------

    // The joint is located at the global origin.  Its kinematic constraints will
    // enforce orthogonality of the associated cross.

    auto ujoint = std::make_shared<ChLinkUniversal>();
    system.AddLink(ujoint);
    ujoint->Initialize(shaft_1, shaft_2, ChFrame<>(ChVector<>(0, 0, 0), rot));

    // Create the Irrlicht application
    // -------------------------------

    ChIrrApp application(&system, L"ChBodyAuxRef demo", core::dimension2d<u32>(800, 600), false, true);
    application.AddTypicalLogo();
    application.AddTypicalSky();
    application.AddTypicalLights();
    application.AddTypicalCamera(core::vector3df(3, 1, -1.5));

    application.AssetBindAll();
    application.AssetUpdateAll();

    // Simulation loop
    application.SetTimestep(0.001);

    int frame = 0;

    while (application.GetDevice()->run()) {
        application.BeginScene();
        application.DrawAll();
        application.DoStep();
        application.EndScene();
        frame++;

        if (frame % 20 == 0) {
            // Output the shaft angular velocities at the current time
            double omega_1 = shaft_1->GetWvel_loc().z();
            double omega_2 = shaft_2->GetWvel_loc().z();
            GetLog() << system.GetChTime() << "   " << omega_1 << "   " << omega_2 << "\n";
        }
    }

    return 0;
}

File: src/demos/python/irrlicht/demo_IRR_ujoint.py
#------------------------------------------------------------------------------
# Name:        pychrono example
# Purpose:
#
# Author:      Lijing Yang
#
# Created:     6/10/2020
# Copyright:   (c) ProjectChrono 2019
#------------------------------------------------------------------------------


import pychrono.core as chrono
import pychrono.irrlicht as chronoirr
import math

print ("Example: demonstration of a universal joint")

# The path to the Chrono data directory containing various assets (meshes, textures, data files)
# is automatically set, relative to the default location of this demo.
# If running from a different directory, you must change the path to the data directory with: 
#chrono.SetChronoDataPath('relative/path/to/data/directory/')

# ---------------------------------------------------------------------
#
#  Create the simulation system and add items
#

mysystem      = chrono.ChSystemNSC()
mysystem.Set_G_acc(chrono.ChVectorD(0, 0, 0))

# Set the half-length of the two shafts
hl = 2
 
# Set the bend angle between the two shafts (positive rotation
# about the global X axis)
angle = math.pi / 6.
cosa  = math.cos(angle)
sina  = math.sin(angle)
rot   = chrono.Q_from_AngX(angle) 

# Create the ground body
# ----------------------
ground = chrono.ChBody()
ground.SetIdentifier(-1)
ground.SetBodyFixed(True)
ground.SetCollide(False)
mysystem.Add(ground)

# attach visualization assets to represent the revolute and cylindrical
# joints that connect the two shafts to ground
cyl_1 = chrono.ChCylinderShape()
cyl_1.GetCylinderGeometry().p1 = chrono.ChVectorD(0, 0, -hl - 0.2)
cyl_1.GetCylinderGeometry().p2 = chrono.ChVectorD(0, 0, -hl + 0.2)
cyl_1.GetCylinderGeometry().rad = 0.3
ground.AddAsset(cyl_1)

cyl_2 = chrono.ChCylinderShape()
cyl_2.GetCylinderGeometry().p1 = chrono.ChVectorD(0, -(hl - 0.2) * sina, (hl - 0.2) * cosa)
cyl_2.GetCylinderGeometry().p2 = chrono.ChVectorD(0, -(hl + 0.2) * sina, (hl + 0.2) * cosa)
cyl_2.GetCylinderGeometry().rad = 0.3
ground.AddAsset(cyl_2)


# Create the first shaft body
# ---------------------------
shaft_1 = chrono.ChBody()
mysystem.AddBody(shaft_1)
shaft_1.SetIdentifier(1)
shaft_1.SetBodyFixed(False)
shaft_1.SetCollide(False)
shaft_1.SetMass(1)
shaft_1.SetInertiaXX(chrono.ChVectorD(1, 1, 0.2))
shaft_1.SetPos(chrono.ChVectorD(0, 0, -hl))
shaft_1.SetRot(chrono.ChQuaternionD(1, 0, 0, 0))

# Add visualization assets to represent the shaft (a box) and the arm of the
# universal joint's cross associated with this shaft (a cylinder)
box_1 = chrono.ChBoxShape()
box_1.GetBoxGeometry().Size = chrono.ChVectorD(0.15, 0.15, 0.9 * hl)
shaft_1.AddAsset(box_1)

cyl_2 = chrono.ChCylinderShape()
cyl_2.GetCylinderGeometry().p1 = chrono.ChVectorD(-0.2, 0, hl)
cyl_2.GetCylinderGeometry().p2 = chrono.ChVectorD(0.2, 0, hl)
cyl_2.GetCylinderGeometry().rad = 0.05
shaft_1.AddAsset(cyl_2)

col = chrono.ChColorAsset()
col.SetColor(chrono.ChColor(0.9, 0.4, 0.1))
shaft_1.AddAsset(col)

# Create the second shaft body
# ----------------------------

# The second shaft is identical to the first one, but initialized at an angle
# equal to the specified bend angle.

shaft_2 = chrono.ChBody()
mysystem.AddBody(shaft_2)
shaft_2.SetIdentifier(1)
shaft_2.SetBodyFixed(False)
shaft_2.SetCollide(False)
shaft_2.SetMass(1)
shaft_2.SetInertiaXX(chrono.ChVectorD(1, 1, 0.2))
shaft_2.SetPos(chrono.ChVectorD(0, -hl * sina, hl * cosa))
shaft_2.SetRot(rot)

# Add visualization assets to represent the shaft (a box) and the arm of the
# universal joint's cross associated with this shaft (a cylinder)
box_1 = chrono.ChBoxShape()
box_1.GetBoxGeometry().Size = chrono.ChVectorD(0.15, 0.15, 0.9 * hl)
shaft_2.AddAsset(box_1)

cyl_2 = chrono.ChCylinderShape()
cyl_2.GetCylinderGeometry().p1 = chrono.ChVectorD(0, -0.2, -hl)
cyl_2.GetCylinderGeometry().p2 = chrono.ChVectorD(0, 0.2, -hl)
cyl_2.GetCylinderGeometry().rad = 0.05
shaft_2.AddAsset(cyl_2)

col = chrono.ChColorAsset()
col.SetColor(chrono.ChColor(0.2, 0.4, 0.8))
shaft_2.AddAsset(col)

# Connect the first shaft to ground
# ---------------------------------

# Use a rotational motor to impose both the revolute joint constraints, as well
# as constant angular velocity. Here, we drive the motor angle with a ramp function.
# Alternatively, we could use a ChLinkMotorAngularSpeed with constant speed.
# The joint is located at the origin of the first shaft.
motor = chrono.ChLinkMotorRotationAngle()
motor.Initialize(ground,
                 shaft_1,
                 chrono.ChFrameD(chrono.ChVectorD(0, 0, -hl), chrono.ChQuaternionD(1, 0, 0, 0)))
motor.SetAngleFunction(chrono.ChFunction_Ramp(0, 1))
mysystem.AddLink(motor)

# Connect the second shaft to ground through a cylindrical joint
# --------------------------------------------------------------

# Use a cylindrical joint so that we do not have redundant constraints
# (note that, technically Chrono could deal with a revolute joint here).
# the joint is located at the origin of the second shaft.

cyljoint = chrono.ChLinkLockCylindrical()
mysystem.AddLink(cyljoint)
cyljoint.Initialize(ground, 
                    shaft_2,
                    chrono.ChCoordsysD(chrono.ChVectorD(0, -hl * sina, hl * cosa), rot))

# Connect the two shafts through a universal joint
# ------------------------------------------------

# The joint is located at the global origin.  Its kinematic constraints will
# enforce orthogonality of the associated cross.

ujoint = chrono.ChLinkUniversal()
mysystem.AddLink(ujoint)
ujoint.Initialize(shaft_1,
                  shaft_2,
                  chrono.ChFrameD(chrono.ChVectorD(0, 0, 0), rot))

# ---------------------------------------------------------------------
#
#  Create an Irrlicht application to visualize the system
#

myapplication = chronoirr.ChIrrApp(mysystem, 'PyChrono example: universal joint', chronoirr.dimension2du(1024,768))

myapplication.AddTypicalSky()
myapplication.AddTypicalLogo(chrono.GetChronoDataFile('logo_pychrono_alpha.png'))
myapplication.AddTypicalCamera(chronoirr.vector3df(3, 1, -1.5))
myapplication.AddTypicalLights()

# ==IMPORTANT!== Use this function for adding a ChIrrNodeAsset to all items
# in the system. These ChIrrNodeAsset assets are 'proxies' to the Irrlicht meshes.
# If you need a finer control on which item really needs a visualization proxy in
# Irrlicht, just use application.AssetBind(myitem) on a per-item basis.

myapplication.AssetBindAll()

# ==IMPORTANT!== Use this function for 'converting' into Irrlicht meshes the assets
# that you added to the bodies into 3D shapes, they can be visualized by Irrlicht!

myapplication.AssetUpdateAll()


# ---------------------------------------------------------------------
#
#  Run the simulation
#
myapplication.SetTimestep(0.001)
myapplication.SetTryRealtime(True)
frame = 0

while(myapplication.GetDevice().run()):
    myapplication.BeginScene()
    myapplication.DrawAll()
    myapplication.DoStep()
    myapplication.EndScene()
    frame += 1

    if frame % 20 == 0:
        omega_1 = shaft_1.GetWvel_loc().z
        omega_2 = shaft_2.GetWvel_loc().z
        print('{:.4}'.format(str(mysystem.GetChTime())), '{:.6}'.format(str(omega_1)), '{:.6}'.format(str(omega_2)))






